Feature/vuex gateway init

kimlisa opened the pull request:
closes #22

kimlisa committed:
Set up vuex store module for app-gateway-init

kimlisa committed:
Move generic types to root-types

kimlisa committed:
Remove comment and fix imports on mutation.spec

kimlisa committed:
Add code coverage enabling in jest

faraonc commented on package.json:
```diff
@@ -16,12 +16,14 @@
     "vue": "^2.6.10",
     "vue-class-component": "^7.0.2",
     "vue-i18n": "^8.0.0",
     "vue-property-decorator": "^8.1.0",
```
8.1.1 is out

kimlisa replied:
updated

faraonc commented on src/store/README.md:
```diff
@@ -0,0 +1,106 @@
+# Store Guide
```
We should reference this README.md in the project's parent directory README.md.

faraonc replied:
Is this resolved already?

faraonc commented on src/store/README.md:
```diff
@@ -0,0 +1,106 @@
+### actions.ts
+Contains exported defined action methods for the store module.
+
+The difference between `mutations` and `actions` is that `actions` contains the heavy logic, 
```
https://vuex.vuejs.org/guide/actions.html 
Not necessarily, better to rephrase this as `actions` is the business logic and can commit multiple `mutations` at a time. `mutations` may have logic dealing with mutating the state.

faraonc replied:
Is this resolved or am I seeing old commits?

faraonc commented on src/store/modules/app-gateway-svc/actions.ts:
```diff
@@ -0,0 +1,55 @@
+  if (hostname.length === 0) {
+    // TODO not sure what to do here
+    console.error('setNewClient: VUE_APP_HOST_NAME env var not loaded');
+  }
```
I think we can use truthy here.
```ts
const hostname: string = process.env.VUE_APP_HOST_NAME;

if (!hostname) {
    // TODO not sure what to do here
    console.error('setNewClient: VUE_APP_HOST_NAME env var not loaded');
    // route to error 50X page
  }
```

faraonc replied:
We can probably removed the alternate type `undefined` as well.

kimlisa replied:
it gives type error because the env var may not exist which is than "undefined" and we cant store undefined as strings.

faraonc replied:
```ts
 const hostname: string | undefined = process.env.VUE_APP_HOST_NAME;
   if (!hostname) {
    // TODO not sure what to do here
    console.error('setNewClient: VUE_APP_HOST_NAME env var not loaded');
  // route to error 500 page
  }
```

faraonc replied:
ooppppps
```ts
const hostname: string = process.env.VUE_APP_HOST_NAME? '' : address;
 if (!hostname) {
    // TODO not sure what to do here
    console.error('setNewClient: VUE_APP_HOST_NAME env var not loaded');
  // route to error 500 page
  }
```

faraonc commented on src/store/modules/app-gateway-svc/actions.ts:
```diff
@@ -0,0 +1,55 @@
+  commit(mutation.SET_GRPC_CLIENT, client);
+};
+
+export const setNewHttpHeader = ({ state, commit }: ActionContext): any => {
```
I think this should be renamed to `setAuthHeader` because it is only dealing with authentication/authorization.

kimlisa replied:
i renamed action to `initAuthHeader` and renamed mutation to `setAuthHeader`

faraonc commented on src/store/modules/app-gateway-svc/actions.ts:
```diff
@@ -0,0 +1,55 @@
+    );
+    authType = consts.BASIC_AUTH;
+  }
+
```
Minimize mutation and alternate types as much as possible.

I think this can work.
```ts
const isAuthToken: boolean = window.localStorage.getItem(consts.LOCAL_STORAGE_TOKEN_KEY)? true : false;

const token: string = isAuthToken? window.localStorage.getItem(consts.LOCAL_STORAGE_TOKEN_KEY) : window.btoa(
      `${process.env.VUE_APP_DUMMY_EMAIL}:${process.env.VUE_APP_DUMMY_PASSWORD}`,
    );

const authType: string  = isAuthToken?  consts.USER_AUTH :  consts.USER_AUTH
```

kimlisa replied:
```window.localStorage.getItem(consts.LOCAL_STORAGE_TOKEN_KEY)``` does not return a boolean, it returns type "string or null". Typescript is really strict

kimlisa replied:
strangely, this works 
```
const isAuthToken: boolean = !!window.localStorage.getItem(consts.LOCAL_STORAGE_TOKEN_KEY);
``` 
Note the double exclamation mark. Not sure how that works right of the bat

kimlisa replied:
Do you want me to use the double exclamation mark point?

kimlisa replied:
Actually, to enforce a boolean, i can do this

```
let isAuthToken: boolean = window.localStorage.getItem(consts.LOCAL_STORAGE_TOKEN_KEY) === null;
```

faraonc replied:
Try not to use `let` keyword as much as possible is the best practice.

kimlisa replied:
I am open for more feedback after you take a look at the new block. I had to put in more cases.

faraonc commented on src/store/modules/app-gateway-svc/states.ts:
```diff
@@ -0,0 +1,7 @@
+// eslint-disable-next-line import/prefer-default-export
+export const state: ModuleState = {
+  grpcClient: undefined as any,
+  httpHeader: undefined as any,
```
why `undefined as any`? We should make them as optional properties with `?`

faraonc replied:
I do not understand how `states.ts` and `types.ts`. Why do we need both?

faraonc replied:
We can get rid of `types.ts`.

Why not just include the interface here to reduce the file to manage. The boiler plating is getting confusing.

```ts
import { AppGatewayServiceClient, grpc } from '../../root-types';

export interface ModuleState {
  grpcClient?: AppGatewayServiceClient;
  httpHeader?: grpc.Metadata;
}

export const state: ModuleState = {};
```

kimlisa replied:
> why `undefined as any`? We should make them as optional properties with `?`

Because I have to initialize it with something and type null or undefined is not a valid type for the type I defined it as. So `undefined as any` works for any type.

I can't make them optional, it has to exist already for my IDE to recognize it because I make references to this state in actions or other files.

kimlisa replied:
> I do not understand how `states.ts` and `types.ts`. Why do we need both?

That's because this module is very small, so it might be hard to see the difference. This module's only purpose is to initialize client and http header when app.vue starts (atm).

`state.ts` only contain state of the module, `types.ts` includes any interfaces and other typings that the module will need.

kimlisa replied:
> We can get rid of `types.ts`.
> 
> Why not just include the interface here to reduce the file to manage. The boiler plating is getting confusing.
> 
> ```ts
> import { AppGatewayServiceClient, grpc } from '../../root-types';
> 
> export interface ModuleState {
>   grpcClient?: AppGatewayServiceClient;
>   httpHeader?: grpc.Metadata;
> }
> 
> export const state: ModuleState = {};
> ```

I was thinking of that too. The file could potentially get large, but I agree its confusing.

faraonc replied:
If it is confusing then we should get rid of it?

faraonc commented on src/store/modules/app-gateway-svc/states.ts:
```diff
@@ -0,0 +1,7 @@
+// eslint-disable-next-line import/prefer-default-export
+export const state: ModuleState = {
+  grpcClient: undefined as any,
+  httpHeader: undefined as any,
```
`httpHeader` is more like `authHeader` at the moment. If we have to change it to `httpHeader` later on, then we can do it later on. I think it is more specific and easier to understand.

faraonc commented on src/store/modules/app-gateway-svc/mutations.ts:
```diff
@@ -0,0 +1,15 @@
+export const mutations: MutationTree<ModuleState> = {
+  [mutation.SET_HTTP_HEADER](state, payload: grpc.Metadata) {
+    Object.assign(state, { httpHeader: payload });
+    // state.httpHeader = payload;
```
We should probably rename the `state` param to maybe `newState` ?

kimlisa replied:
It's either a convention to use `state` or it actually refers to the `export const state` object we defined in `states.ts`. I know in `actions.ts`, `state` definitely refers to the defined state object and can't change the name.

faraonc commented on src/store/modules/app-gateway-svc/index.ts:
```diff
@@ -0,0 +1,17 @@
+// namespace name is set to the module name: appGateway
+const namespaced: boolean = true;
+
+// eslint-disable-next-line import/prefer-default-export
```
Maybe?
```ts
const appGateway: Module<ModuleState, RootState> = {
  namespaced,
  state,
  actions,
  mutations,
};

export default appGateway;
```

kimlisa replied:
I read that we should avoid using export defaults: https://basarat.gitbooks.io/typescript/docs/tips/defaultIsBad.html

faraonc replied:
the eslint says something different. i think people have different opinions. if we are going to use a linter, might as well stick to it?

faraonc reviewed (commented):
Added my thoughts for you to consider.

kimlisa committed:
Make change based on code review

- Update eslintrc to ignore single exports and increased max-len to 120
- Changed appGateway store module to gateway module
- Collapsed types.ts files into roots types.ts
- Changed setHttpHeader to authHeader
- Corrected explanation about actions in store/README.md

faraonc commented on src/store/modules/gateway/actions.ts:
```diff
@@ -0,0 +1,59 @@
+}
+
+export const setNewClient = ({ state, commit }: ActionContext): any => {
+  const hostname: string = process.env.VUE_APP_HOST_NAME == null ? '' : process.env.VUE_APP_HOST_NAME;
```
Another JS feature, you prolly interested with is `||`.

`const hostname: string = process.env.VUE_APP_HOST_NAME || '';`

faraonc replied:
I think `||` and `!!` exist for reasons like this. I think this a language feature that we should take advantage.

faraonc replied:
Normally, there is another reviewer to settle differences. Maybe we can get opinion from @yeseulan? She worked at Disney as a frontend dev using Angular.

kimlisa replied:
Alright. You guys have industry level experience I don't.

faraonc commented on src/store/modules/gateway/actions.ts:
```diff
@@ -0,0 +1,59 @@
+  if (token == null) {
+    const dummyEmail: string | undefined = process.env.VUE_APP_DUMMY_EMAIL;
+    const dummyPassword: string | undefined = process.env.VUE_APP_DUMMY_PASSWORD;
+    if (dummyEmail == null && dummyPassword == null) {
```
Did you mean `dummyEmail == null || dummyPassword == null`?

kimlisa replied:
yes

faraonc commented on package.json:
```diff
@@ -16,12 +16,14 @@
   "dependencies": {
```
running `npm run serve` gets this error.
```
ERROR in /Users/faraonc/WebstormProjects/hwsc-frontend/node_modules/@hwsc/hwsc-api-blocks-lisa-test/protobuf/hwsc-app-gateway-svc/app/hwsc-app-gateway-svc_pb.d.ts
1:23 Could not find a declaration file for module 'google-protobuf'. '/Users/faraonc/WebstormProjects/hwsc-frontend/node_modules/google-protobuf/google-protobuf.js' implicitly has an 'any' type.
  Try `npm install @types/google-protobuf` if it exists or add a new declaration (.d.ts) file containing `declare module 'google-protobuf';`
  > 1 | import * as jspb from "google-protobuf"
      |                       ^
    2 | 
    3 | import * as protobuf_hwsc$user$svc_user_hwsc$user$svc_pb from '../../../protobuf/hwsc-user-svc/user/hwsc-user-svc_pb';
    4 | import * as protobuf_lib_user_pb from '../../../protobuf/lib/user_pb';
```

Do we need ` npm install @types/google-protobuf`?

faraonc replied:
`npm install @types/google-protobuf` seems to fix the issue but still
```
faraonc@~/WebstormProjects/hwsc-frontend (feature/vuex-gateway-init)$ npm run serve

> hwsc-frontend@0.0.1 serve /Users/faraonc/WebstormProjects/hwsc-frontend
> vue-cli-service serve

 INFO  Starting development server...
Starting type checking service...
Using 1 worker with 2048MB memory limit
 98% after emitting CopyPlugin                                                    

 WARNING  Compiled with 2 warnings                                                                                                                                                                                                                               12:55:36 AM

Module Warning (from ./node_modules/eslint-loader/index.js):
error: Line 11 exceeds the maximum line length of 120 (max-len) at src/components/HelloWorld.vue:11:1:
   9 |     <h3>Installed CLI Plugins</h3>
  10 |     <ul>
> 11 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-babel" target="_blank" rel="noopener">babel</a></li>
     | ^
  12 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-typescript" target="_blank" rel="noopener">typescript</a></li>
  13 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-eslint" target="_blank" rel="noopener">eslint</a></li>
  14 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-unit-jest" target="_blank" rel="noopener">unit-jest</a></li>


error: Line 12 exceeds the maximum line length of 120 (max-len) at src/components/HelloWorld.vue:12:1:
  10 |     <ul>
  11 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-babel" target="_blank" rel="noopener">babel</a></li>
> 12 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-typescript" target="_blank" rel="noopener">typescript</a></li>
     | ^
  13 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-eslint" target="_blank" rel="noopener">eslint</a></li>
  14 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-unit-jest" target="_blank" rel="noopener">unit-jest</a></li>
  15 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-e2e-cypress" target="_blank" rel="noopener">e2e-cypress</a></li>


error: Line 13 exceeds the maximum line length of 120 (max-len) at src/components/HelloWorld.vue:13:1:
  11 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-babel" target="_blank" rel="noopener">babel</a></li>
  12 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-typescript" target="_blank" rel="noopener">typescript</a></li>
> 13 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-eslint" target="_blank" rel="noopener">eslint</a></li>
     | ^
  14 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-unit-jest" target="_blank" rel="noopener">unit-jest</a></li>
  15 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-e2e-cypress" target="_blank" rel="noopener">e2e-cypress</a></li>
  16 |     </ul>


error: Line 14 exceeds the maximum line length of 120 (max-len) at src/components/HelloWorld.vue:14:1:
  12 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-typescript" target="_blank" rel="noopener">typescript</a></li>
  13 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-eslint" target="_blank" rel="noopener">eslint</a></li>
> 14 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-unit-jest" target="_blank" rel="noopener">unit-jest</a></li>
     | ^
  15 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-e2e-cypress" target="_blank" rel="noopener">e2e-cypress</a></li>
  16 |     </ul>
  17 |     <h3>Essential Links</h3>


error: Line 15 exceeds the maximum line length of 120 (max-len) at src/components/HelloWorld.vue:15:1:
  13 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-eslint" target="_blank" rel="noopener">eslint</a></li>
  14 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-unit-jest" target="_blank" rel="noopener">unit-jest</a></li>
> 15 |       <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-e2e-cypress" target="_blank" rel="noopener">e2e-cypress</a></li>
     | ^
  16 |     </ul>
  17 |     <h3>Essential Links</h3>
  18 |     <ul>


error: Line 29 exceeds the maximum line length of 120 (max-len) at src/components/HelloWorld.vue:29:1:
  27 |       <li><a href="https://router.vuejs.org" target="_blank" rel="noopener">vue-router</a></li>
  28 |       <li><a href="https://vuex.vuejs.org" target="_blank" rel="noopener">vuex</a></li>
> 29 |       <li><a href="https://github.com/vuejs/vue-devtools#vue-devtools" target="_blank" rel="noopener">vue-devtools</a></li>
     | ^
  30 |       <li><a href="https://vue-loader.vuejs.org" target="_blank" rel="noopener">vue-loader</a></li>
  31 |       <li><a href="https://github.com/vuejs/awesome-vue" target="_blank" rel="noopener">awesome-vue</a></li>
  32 |     </ul>


6 errors found.

Module Warning (from ./node_modules/eslint-loader/index.js):
error: Missing semicolon (semi) at src/components/HelloI18n.vue:11:56:
   9 | 
  10 | <script lang="ts">
> 11 | import { Component, Vue } from 'vue-property-decorator'
     |                                                        ^
  12 | 
  13 | @Component
  14 | export default class HelloI18n extends Vue {


1 error found.
1 error potentially fixable with the `--fix` option.

You may use special comments to disable some warnings.
Use // eslint-disable-next-line to ignore the next line.
Use /* eslint-disable */ to ignore all warnings in a file.                                                                                                                                                                                    Type checking in progress...  

  App running at:
  - Local:   http://localhost:8080/ 
  - Network: http://10.0.0.182:8080/

  Note that the development build is not optimized.
  To create a production build, run npm run build.

No type errors found
Version: typescript 3.4.5
Time: 10798ms
```

faraonc replied:
I did hack this line `responseDeserializeFn: (bytes: Uint8Array) => Response);` just to try it out.

kimlisa replied:
I was going to fix this on another branch. But I fixed it anyways. The only lint errors now are `(bytes: Uint8Array)` related.

faraonc reviewed (changes requested):
Some npm and lint issues

faraonc commented on src/store/modules/gateway/types.ts:
```diff
@@ -0,0 +1,6 @@
+export interface ModuleState {
+  grpcClient: AppGatewayServiceClient;
+  httpHeader: grpc.Metadata;
+}
```
Can we just get rid of this file and move the interface to `state.ts` just like `ActionContext` in `actions.ts`? This is just an extra file.

kimlisa replied:
I meant to delete this file. It's already moved to root `types.ts`

faraonc commented on src/store/modules/gateway/actions.ts:
```diff
@@ -0,0 +1,59 @@
+}
+
+export const setNewClient = ({ state, commit }: ActionContext): any => {
+  const hostname: string = process.env.VUE_APP_HOST_NAME == null ? '' : process.env.VUE_APP_HOST_NAME;
```
We need to update `VUE_APP_HOST_NAME` in our slack channel config file

faraonc replied:
Are you using this value `http://localhost:50056`?

kimlisa replied:
Updated. Yes its the envoy proxy port

faraonc reviewed (changes requested):
I suggest we find another reviewer to settle code review arguments.

faraonc commented on tests/unit/store/gateway/mutations.spec.ts:
```diff
@@ -0,0 +1,30 @@
+  });
+});
+
+// https://github.com/grpc/grpc-web/pull/540
```
We need a task to track this.

kimlisa replied:
https://github.com/hwsc-org/hwsc-frontend/issues/27

kimlisa committed:
Fix per code review

- Replaced test hwsc-api-blocks with official release
- Modified eslint to ignore long urls and strings
- Deleted types.ts file in store/module/gateway directory
- Edited javascript to use shorthands

faraonc reviewed (approved):
thank you and great work. TU
